### PR TITLE
Implement minimal animation graphs in order to support animation blending.

### PR DESCRIPTION
**Pull request by pcwalton**
# Objective

This is a minimum viable implementation of https://github.com/bevyengine/rfcs/pull/51. It allows Bevy to support animation blending.

## Solution

This is just to start the conversation.

### CODE REVIEW

File: crates/bevy_animation/src/graph.rs
Comment (james7132) [inline crates/bevy_animation/src/graph.rs:221]: This seems to drop the parallelization that is currently supported, and may be a significant performance regression for scenes with a large number of animated rigs.

File: crates/bevy_animation/src/graph.rs
Comment (james7132) [inline crates/bevy_animation/src/graph.rs:168]: Each time `apply_animation` is called, the query for each bone within the hierarchy is redone, which involves a hierarchy traversal and multiple string comparisons. This works, but the total overhead scales with the number of clips in the graph. This isn't a significant departure from what we currently have, so I don't want to block on this, but we need to address this sooner rather than later.

File: crates/bevy_animation/src/graph.rs
Comment (james7132) [inline crates/bevy_animation/src/graph.rs:74]: These lifetimes definitely need to be addressed before we merge this. This is a little ridiculous.

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_animation/src/graph.rs
Comment (mbrea-c) [reply crates/bevy_animation/src/graph.rs:221]: Is there any reason we cannot just use `.par_iter_mut().for_each` here? Seems like it would work.

File: crates/bevy_animation/src/graph.rs
Comment (mbrea-c) [reply crates/bevy_animation/src/graph.rs:74]: Can these not share the same lifetime bound? See for example [here](https://github.com/mbrea-c/bevy_animation_graph/blob/master/crates/bevy_animation_graph/src/core/context/pass_context.rs#L23) and [here](https://github.com/mbrea-c/bevy_animation_graph/blob/master/crates/bevy_animation_graph/src/core/context/system_resources.rs#L12).

Comment (NthTensor) [reply crates/bevy_animation/src/graph.rs:74]: I was able to appease the borrow-checker with the following:
```rust
struct AnimationContext<'l, 'w, 's, 'c> {
    animation_graph: &'l mut AnimationGraph,
    time: &'l Time,
    root_entity: Entity,
    animation_clips: &'l Assets<AnimationClip>,
    parent: Option<&'w Parent>,
    parents: &'l Query<
        'w,
        's,
        (
            Has<AnimationPlayer>,
            Has<AnimationGraph>,
            Option<&'c Parent>,
        ),
    >,
    children: &'l Query<'w, 's, &'c Children>,
    names: &'l Query<'w, 's, &'c Name>,
    transforms: &'l Query<'w, 's, &'c mut Transform>,
    morph_weights: &'l Query<'w, 's, &'c mut MorphWeights>,
}

pub fn evaluate_animation_graphs<'w>(
    time: Res<Time>,
    animation_clips: Res<Assets<AnimationClip>>,
    children: Query<&'w Children>,
    names: Query<&'w Name>,
    transforms: Query<&'w mut Transform>,
    morph_weights: Query<&'w mut MorphWeights>,
    parents: Query<(
        Has<AnimationPlayer>,
        Has<AnimationGraph>,
        Option<&'w Parent>,
    )>,
    mut animation_graphs: Query<(Entity, Option<&'w Parent>, &'w mut AnimationGraph)>,
)
```
I don't *think* it is possible to reduce this further. Not currently sure what `AnimationContext` is for but I trust the process.

File: crates/bevy_animation/src/graph.rs
Comment (NthTensor) [reply crates/bevy_animation/src/graph.rs:221]: It does look like `par_iter` would bring this in line with how things are currently dispatched in `animation_player`. Same exact query more or less.

**Comment by pcwalton** [comment]
I'm closing this as there will need to be a *lot* of changes.

**Pull request closed without merging**
